### PR TITLE
[FLINK-28868][connector/hbase] Migrate HBase table connector to the new LookupFunction interface

### PR DESCRIPTION
**Pull request by PatrickRen**
## What is the purpose of the change

This PR migrates JDBC table connector to the new LookupFunction interface

## Brief change log

- Migrate JDBC table connector to the new LookupFunction interface

## Verifying this change

This change is already covered by existing HBase connector tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): (yes / **no**)
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: (yes / **no**)
  - The serializers: (yes / **no** / don't know)
  - The runtime per-record code paths (performance sensitive): (yes / **no** / don't know)
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: (yes / **no** / don't know)
  - The S3 file system connector: (yes / **no** / don't know)

## Documentation

  - Does this pull request introduce a new feature? (yes / **no**)
  - If yes, how is the feature documented? (**not applicable** / docs / JavaDocs / not documented)

### CODE REVIEW

File: flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/source/HBaseDynamicTableSource.java [hunk 1]
```diff
@@ -38,23 +39,19 @@ public HBaseDynamicTableSource(
...
     }
 
     @Override
     public DynamicTableSource copy() {
         return new HBaseDynamicTableSource(
-                conf, tableName, hbaseSchema, nullStringLiteral, lookupOptions);
+                conf, tableName, hbaseSchema, nullStringLiteral, maxRetryTimes, cache);
     }
 
     @Override
```
Comment (leonardBang) [inline flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/source/HBaseDynamicTableSource.java:53]: Add missed `equals()` and `hashcode()` methods?

File: flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1DynamicTableFactory.java [hunk 2]
```diff
@@ -78,13 +81,31 @@ public DynamicTableSource createDynamicTableSource(Context context) {
         String nullStringLiteral = tableOptions.get(NULL_STRING_LITERAL);
         HBaseTableSchema hbaseSchema =
                 HBaseTableSchema.fromDataType(context.getPhysicalRowDataType());
+        LookupCache cache = null;
+
+        // Backward compatible to legacy caching options
+        if (tableOptions.get(LOOKUP_CACHE_MAX_ROWS) > 0
```
Comment (leonardBang) [inline flink-connectors/flink-connector-hbase-1.4/src/main/java/org/apache/flink/connector/hbase1/HBase1DynamicTableFactory.java:87]: Could we use fallbackKeys way?

File: flink-connectors/flink-connector-hbase-1.4/src/test/java/org/apache/flink/connector/hbase1/HBaseDynamicTableFactoryTest.java [hunk 3]
```diff
@@ -141,6 +143,32 @@ public void testTableSourceFactory() {
                 hbaseSchema.getQualifierDataTypes("f4"));
     }
 
+    @Test
+    public void testCacheOptions() {
```
Comment (leonardBang) [inline flink-connectors/flink-connector-hbase-1.4/src/test/java/org/apache/flink/connector/hbase1/HBaseDynamicTableFactoryTest.java:147]:
```suggestion
    public void testLookupOptions() {
```
`lookup.max-retries` is not related to cachiung.

File: flink-connectors/flink-connector-hbase-2.2/src/test/java/org/apache/flink/connector/hbase2/HBaseDynamicTableFactoryTest.java [hunk 4]
```diff
@@ -144,6 +146,32 @@ public void testTableSourceFactory() {
                 hbaseSchema.getQualifierDataTypes("f4"));
     }
 
+    @Test
+    public void testCacheOptions() {
```
Comment (leonardBang) [inline flink-connectors/flink-connector-hbase-2.2/src/test/java/org/apache/flink/connector/hbase2/HBaseDynamicTableFactoryTest.java:150]:
```suggestion
    public void testLookupOptions() {
```

File: flink-connectors/flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/table/HBaseConnectorOptions.java [hunk 5]
```diff
@@ -93,6 +94,8 @@ public class HBaseConnectorOptions {
                     .defaultValue(false)
                     .withDescription("whether to set async lookup.");
 
+    /** @deprecated Please use {@link LookupOptions#PARTIAL_CACHE_MAX_ROWS} instead. */
+    @Deprecated
     public static final ConfigOption<Long> LOOKUP_CACHE_MAX_ROWS =
```
Comment (leonardBang) [inline flink-connectors/flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/table/HBaseConnectorOptions.java:99]: add fallbackKeys

File: flink-connectors/flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/source/HBaseRowDataLookupFunction.java [hunk 6]
```diff
@@ -68,54 +65,37 @@ public class HBaseRowDataLookupFunction extends TableFunction<RowData> {
...
-                return;
-            }
-        }
+    @Override
+    public Collection<RowData> lookup(RowData keyRow) throws IOException {
         for (int retry = 0; retry <= maxRetryTimes; retry++) {
             try {
                 // fetch result
-                Get get = serde.createGet(rowKey);
+                Get get = serde.createGet(((GenericRowData) keyRow).getField(0));
```
Comment (leonardBang) [inline flink-connectors/flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/source/HBaseRowDataLookupFunction.java:94]: `GenericRowData` is only one kind of implementation of `RowData`, the framework can ensure the keyRow is `GenericRowData` ?

**Review by leonardBang — COMMENTED** [review]
Thanks @PatrickRen for the contribution, I left some comments

File: flink-connectors/flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/source/HBaseRowDataLookupFunction.java (hunk 6 above)
Comment (PatrickRen) [reply flink-connectors/flink-connector-hbase-base/src/main/java/org/apache/flink/connector/hbase/source/HBaseRowDataLookupFunction.java:94]: The implementation of LookupFunction will pass a GenericRowData as key row and it's safe to cast for now. I added a TODO here to update the logic once we improve the LookupFunction in the future.

**Review by leonardBang — APPROVED** [review]
Thanks @PatrickRen  for the contribution, LGTM

**Comment by PatrickRen** [comment]
CI passed except an unrelated case tracked by FLINK-28767: https://dev.azure.com/renqs/Apache%20Flink/_build/results?buildId=388&view=results

**Pull request closed without merging**